### PR TITLE
Apply blobless_fetch to extra_refs

### PR DESCRIPTION
**Pull request by airbornepony**
Set BloblessFetch in ExtraRefs based on the DecorationConfig. This ensures that periodic jobs (which only specify ExtraRefs) and any presubmit, etc. jobs that happen to specify extra_refs use blobless clones consistently.

Setting BloblessFetch in ExtraRefs takes preference over the default config. This is done by making that field a pointer to boolean; only nil values are replaced by the default.

### CODE REVIEW

File: prow/pjutil/pjutil_test.go
Comment (listx) [inline prow/pjutil/pjutil_test.go:1229]: Nit: I would prefer `extraRef` over `er` for readability. The `er` looks too close to `err` for me. Perhaps even `defaultedExtraRef` is another good candidate, because you're expecting the BloblessFetch in the DecorationConfig to override the false value in ExtraRefs.

File: prow/pjutil/pjutil_test.go
Comment (listx) [inline prow/pjutil/pjutil_test.go:1222]: How about adding another ExtraRef case with `BloblessFetch: true` here to check that it stays `true` after defaulting?

Another separate case you could add is what happens if BloblessFetch is not specified in ExtraRefs (i.e., does it pick up the `true` setting from DecorationConfig?).

**Review by listx — COMMENTED** [review]
Generally looks good. I have some nits for additional test cases. But also, you say in your commit message

> Since Refs.BloblessFetch is a simple boolean, the DecorationConfig
default cannot be overridden by setting a true or false value in
ExtraRefs. In the unlikely case the default must be overridden, the
DecorationConfig.BloblessFetch field should be used instead.

and I'm not quite grokking what you mean exactly. Is it possible to demonstrate what you mean with test cases here (is it straightforward using existing test functions)? If so, it would be good to add them as test cases.

File: prow/pjutil/pjutil_test.go
Comment (airbornepony) [reply prow/pjutil/pjutil_test.go]: Agree, er is almost err. Hope just 'r' is not confusing. I like to keep loop variables short to maintain readability, and it's Go style.

Also, create a new slice just in case it's not safe to modify the one from the config.JobBase.

File: prow/pjutil/pjutil_test.go
Comment (airbornepony) [reply prow/pjutil/pjutil_test.go]: I've written three tests cases now, demonstrating true and false defaults, and preservation of values without defaults. I hope this clarifies the meaning of my statement in the description.

Comment (listx) [reply prow/pjutil/pjutil_test.go]: I understand the test cases well enough, but your statement in the description is still hard to understand for me. Maybe it's just me, but I don't have a strong opinion on this, so LGTM.

Comment (airbornepony) [reply prow/pjutil/pjutil_test.go]: I've reworded the note in the description. Let me know if you have suggestions.

File: prow/pjutil/pjutil_test.go
Comment (stevekuznetsov) [inline prow/pjutil/pjutil_test.go:1221]:
```suggestion
						BloblessFetch: ptr.To(true),
```

nit: that utility makes it a little easier - the reader no longer has to trust that `trueVar` actually holds `true` ;)

**Comment by stevekuznetsov** [comment]
> Setting BloblessFetch in ExtraRefs will not override the default config

Just wanted to confirm this is how users expect the override to go - usually in Prow the more specific configuration (here one inside of a specific `extra_ref`) would override the more generic one.

**Comment by stevekuznetsov** [comment]
I think that would be necessary to allow us to make the UX more obvious. @listx WDYT?

File: prow/pjutil/pjutil_test.go
Comment (airbornepony) [reply prow/pjutil/pjutil_test.go:1221]: Good point. Rather than import a package not used by Prow, I noticed that pjutil.go already has boolPtr(). Will use that.

Comment (stevekuznetsov) [reply prow/pjutil/pjutil_test.go:1221]: Feel free not to change this - only a nit.

**Comment by listx** [comment]
So currently when we have DecorationConfig and want to override the more specific config (in this case BloblessFetch in ExtraRefs), we would really like to know if the more specific setting was explicitly set to false, explicitly set to true, or was not set at all (missing). Currently the regular `bool` field (and not pointer to bool) prevents us from telling the difference between explicitly-false and missing, because if it is missing the zero value for it is false and that's how it'll get unmarshaled from the YAML.

Moving to a pointer-to-bool sounds right to me.

**Comment by stevekuznetsov** [comment]
>  currently when we have DecorationConfig and want to override the more specific config (in this case BloblessFetch in ExtraRefs)

@listx  I think my question was whether we really wanted that UX - since in Prow we generally make the more specific config win, which is the opposite of what is happening here

**Comment by airbornepony** [comment]
> > currently when we have DecorationConfig and want to override the more specific config (in this case BloblessFetch in ExtraRefs)
> 
> @listx I think my question was whether we really wanted that UX - since in Prow we generally make the more specific config win, which is the opposite of what is happening here

I think we're all in overwhelming agreement. I'm refactoring this now.

File: prow/pjutil/pjutil_test.go
Comment (airbornepony) [reply prow/pjutil/pjutil_test.go]: Done

**Comment by listx** [comment]
> > currently when we have DecorationConfig and want to override the more specific config (in this case BloblessFetch in ExtraRefs)
> 
> @listx I think my question was whether we really wanted that UX - since in Prow we generally make the more specific config win, which is the opposite of what is happening here

I thought we were saying that we only wanted to override the more specific setting only if the user does not explicitly set blobless fetch to false in their more specific config? IOW we would only want to override if this field from generic to specific if the specific setting is missing.

Overriding an explicitly-set specific setting that has blobless clone set to false for a particular ref in a particular job sounds wrong (I don't think anyone proposed this idea in this discussion).

**Comment by airbornepony** [comment]
PTAL

I've refactored (third commit) Refs.BloblessFetch to a bool pointer to make this work intuitively, only replacing unspecified values with the default value.

File: prow/pjutil/pjutil_test.go
Comment (listx) [inline prow/pjutil/pjutil_test.go:1309]: Shouldn't we also check the "default-org" case here (expect BloblessFetch to be nil)?

Comment (airbornepony) [reply prow/pjutil/pjutil_test.go:1309]: nil is the default value of want, and every Refs is always checked.

**Review by BenTheElder — COMMENTED** [review]
LGTM, after the pointer change. Thank you!

**Comment by airbornepony** [comment]
I've added a final commit emending the Refs.BloblessFetch docs to clarify the default behavior.

If there are no other concerns, I'll await LGTMs and unhold this. Thanks everyone!

File: prow/pjutil/pjutil_test.go
Comment (listx) [reply prow/pjutil/pjutil_test.go:1309]: > nil is the default value of want
Ah, because you are relying on the default zero-value of the pointer-to-bool in `var want *bool`.

You're correct that nil is already the value here, but it's odd that we don't actually use the `BloblessFetch: nil` line on 1295. It makes me think if line 1295 should be deleted. But then that would make the test case less explicit, which hurts readability even more.

LGTM code-wise, but style-wise I would add a comment about this somewhere.

**Comment by listx** [comment]
> I've added a final commit emending the Refs.BloblessFetch docs to clarify the default behavior.
> 
> If there are no other concerns, I'll await LGTMs and unhold this. Thanks everyone!

LGTM

File: prow/pjutil/pjutil_test.go
Comment (airbornepony) [reply prow/pjutil/pjutil_test.go:1312]: PTAL. Added the "default-org" case explicitly.

**Review by stevekuznetsov — APPROVED** [review]

**Pull request merged**
